allennlp/modules/seq2seq_encoders/bidirectional_transformer_encoder.py

```diff
@@ -0,0 +1,265 @@
+        nbatches = query.size(0)
+
+        # 1) Do all the linear projections in batch from d_model => h x d_k
+        query, key, value = [layer(x).view(nbatches, -1, self.num_heads, self.d_k).transpose(1, 2)
+                             for layer, x in zip(self.linears, (query, key, value))]
+
+        # 2) Apply attention on all the projected vectors in batch.
+        x, _ = attention(query, key, value, mask=mask, dropout=self.dropout)
+
+        # 3) "Concat" using a view and apply a final linear.
+        x = x.transpose(1, 2).contiguous().view(nbatches, -1, self.num_heads * self.d_k)
+        return self.linears[-1](x)
+
+
+def make_model(num_layers: int = 6,
+               input_size: int = 512, # Attention size
+               hidden_size: int = 2048, # FF layer size
+               heads: int = 8,
+               dropout: float = 0.1,
+               return_all_layers: bool = False) -> TransformerEncoder:
+    "Helper: Construct a model from hyperparameters."
+    attn = MultiHeadedAttention(heads, input_size, dropout)
+    ff = PositionwiseFeedForward(input_size, hidden_size, dropout)
+    model = TransformerEncoder(EncoderLayer(input_size, attn, ff, dropout),
+                               num_layers,
+                               return_all_layers=return_all_layers)
+
+    # Initialize parameters with Glorot / fan_avg.
+    for p in model.parameters():
+        if p.dim() > 1:
+            torch.nn.init.xavier_uniform(p)
+    return model
+
+
+class BidirectionalTransformerEncoder(Seq2SeqEncoder):
+    def __init__(self,
+                 input_dim: int,
+                 hidden_dim: int,
+                 num_layers: int,
+                 dropout: float = 0.1,
+                 input_dropout: float = None,
+                 return_all_layers: bool = False) -> None:
+        super().__init__()
+
+        self._return_all_layers = return_all_layers
+        self.transformer_layers = num_layers
+        self.num_layers = num_layers
+
+        self._forward_transformer = make_model(input_size=input_dim,
+                                               hidden_size=hidden_dim,
+                                               num_layers=num_layers,
+                                               dropout=dropout,
+                                               return_all_layers=return_all_layers)
+        self._backward_transformer = make_model(input_size=input_dim,
+                                                hidden_size=hidden_dim,
+                                                num_layers=num_layers,
+                                                dropout=dropout,
+                                                return_all_layers=return_all_layers)
+        self._position = PositionalEncoding(input_dim)
+
+        self.input_dim = input_dim
+        self.output_dim = 2 * input_dim
+
+        if input_dropout:
+            self._dropout = torch.nn.Dropout(input_dropout)
+        else:
+            self._dropout = lambda x: x
```

Comment:
For my edification, have we found there to be substantial overhead to `torch.nn.Dropout(0.0)`?